rollout/trafficrouting/plugin/plugin.go

```diff
@@ -0,0 +1,102 @@
+package plugin
+
+import (
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/rollout/trafficrouting/plugin/client"
+	"github.com/argoproj/argo-rollouts/rollout/trafficrouting/plugin/rpc"
+	"github.com/argoproj/argo-rollouts/utils/record"
+	"k8s.io/client-go/kubernetes"
+)
+
+const ErrNotImplemented = "not-implemented"
+
+type ReconcilerConfig struct {
+	Rollout    *v1alpha1.Rollout
+	PluginName string
+	Client     kubernetes.Interface
+	Recorder   record.EventRecorder
+}
+
+type Reconciler struct {
+	Rollout    *v1alpha1.Rollout
+	PluginName string
+	Client     kubernetes.Interface
+	Recorder   record.EventRecorder
+	rpc.TrafficRouterPlugin
+}
+
+func NewReconciler(cfg *ReconcilerConfig) (*Reconciler, error) {
+	pluginClient, err := client.GetTrafficPlugin(cfg.PluginName)
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler := &Reconciler{
+		Rollout:             cfg.Rollout,
+		Client:              cfg.Client,
+		Recorder:            cfg.Recorder,
+		PluginName:          cfg.PluginName,
+		TrafficRouterPlugin: pluginClient,
+	}
+	return reconciler, nil
+}
+
+// UpdateHash informs a traffic routing reconciler about new canary, stable, and additionalDestination(s) pod hashes
+func (r *Reconciler) UpdateHash(canaryHash, stableHash string, additionalDestinations ...v1alpha1.WeightDestination) error {
+	err := r.TrafficRouterPlugin.UpdateHash(r.Rollout, canaryHash, stableHash, additionalDestinations)
+	if err.Error() != "" {
```

Comment:
if err is nil we are going to have a NilPointer error here.

Reply:
err will never be nil the rpc code will instantiate it as it's empty type